File: herencia-dto/HerenciaDto-Java-IvanRC/src/main/java/mapper/PersonaMapper.java
package mapper;

import dto.ListaPersonasDto;
import dto.PersonaDto;
import models.Alumno;
import models.Persona;
import models.Profesor;

import java.util.List;

public class PersonaMapper {

    public static PersonaDto toDto(Persona persona){
        PersonaDto personaDto = null;
        if(persona instanceof Profesor){
            personaDto = new PersonaDto(
                    persona.getNombre(),
                    null,
                    ((Profesor) persona).getModulo(),
                    "profesor"
            );
        }else{
            personaDto = new PersonaDto(
                    persona.getNombre(),
                    String.valueOf(((Alumno) persona).getEdad()),
                    null,
                    "alumno"
            );
        }
        return personaDto;
    }

    public static Persona toPersona(PersonaDto personaDto){
        Persona persona = null;
        if(personaDto.getTipo().equals("profesor")){
            persona = new Profesor(
                    personaDto.getNombre(),
                    personaDto.getModulo()
            );
        }else{
            persona = new Alumno(
                    personaDto.getNombre(),
                    Integer.parseInt(personaDto.getEdad())
            );
        }
        return persona;
    }

    public static ListaPersonasDto toPersonasDto(List<Persona> personas){
        return new ListaPersonasDto(
                personas.stream().map(PersonaMapper::toDto).toList()
        );
    }

    public static List<Persona> toPersonas(ListaPersonasDto personasDto){
        return personasDto.getPersonaDtos().stream().map(PersonaMapper::toPersona).toList();
    }

}

File: herencia-dto/HerenciaDto-Kotlin-IvanRC/src/main/kotlin/mapper/PersonaMapper.kt
package mapper

import dto.ListaPersonasDto
import dto.PersonaDto
import exception.PersonaBadRequestException
import models.Alumno
import models.Persona
import models.Profesor
import mu.KotlinLogging

private val logger = KotlinLogging.logger {  }

fun Persona.toDto(): PersonaDto{
    logger.debug { "Se crea una personaDto, a partir de una persona." }
    var personaDto: PersonaDto? = null;
    if(this is Profesor){
        personaDto = PersonaDto(
            nombre = this.nombre,
            modulo = this.modulo,
            tipo = "profesor"
        )
    }else{
        if(this is Alumno) {
            personaDto = PersonaDto(
                nombre = this.nombre,
                edad = this.edad.toString(),
                tipo = "alumno"
            )
        }
    }
    return personaDto?: throw PersonaBadRequestException("Hubo algún error al tratar de crear a una personaDto.")
}

fun PersonaDto.toPersona(): Persona{
    logger.debug { "Se crea una persona, a partir de una personaDto." }
    var persona: Persona? = null
    if(this.tipo == "profesor"){
        persona = Profesor(
            nombre = this.nombre,
            modulo = this.modulo!!
        )
    }else{
        persona = Alumno(
            nombre = this.nombre,
            edad = this.edad!!.toInt()
        )
    }
    return persona
}

fun List<Persona>.toPersonasDto(): ListaPersonasDto{
    logger.debug { "Se crea una lista de personasDto, a partir de una lista de personas." }
    return ListaPersonasDto(
        this.map { it.toDto() }
    )
}

fun ListaPersonasDto.toPersonas(): List<Persona>{
    logger.debug { "Se crea una lista de personas, a partir de una lista de personasDto." }
    return this.personasDto.map { it.toPersona() }
}